[#11]feat: 로그아웃 기능

Yo0oN opened the pull request:
로그아웃 기능 추가

Yo0oN committed:
[#11]feat: 로그아웃 기능

로그아웃 기능 생성

[#11] 로그아웃 기능

Jane096 commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -48,4 +48,15 @@ public HttpStatus loginMember(@RequestBody @Valid LoginDTO loginDTO, HttpSession
+     * @param httpSession
+     * @return HttpStatus
+     */
+    @PostMapping(value = "/logout")
+    public HttpStatus logoutMember(@RequestBody @Valid HttpSession httpSession){
```
여기서 httpSession에 `@Valid`를 걸어주셨는데 어떤 유효성 체크를 하는지 궁금합니다!

Yo0oN replied:
이부분은 손에 익은대로 타이핑하다보니 사용하지 않는 부분도 입력하게 된것같습니다..필요 없는 부분을 다시 한번 검토하고 커밋하겠습니다!

Yo0oN committed:
[#11]feat: 로그아웃 기능

사용하지 않는 @Valid 삭제

[#11] 로그아웃 기능

f-lab-dev commented on src/main/java/com/festa/controller/HostController.java:
```diff
@@ -41,4 +41,15 @@ public HttpStatus loginHost(@RequestBody @Valid LoginDTO loginDTO, HttpSession h
+    @PostMapping(value = "/logout")
+    public HttpStatus logoutHost(@RequestBody HttpSession httpSession){
+        accountsService.logout(httpSession);
+        return HttpStatus.OK;
```
1. 이럴땐 리턴타입을 void로 해주면 리턴해줄 필요가 없어서 더 깔끔할 것 같네요~
2. 아직 Host와 Member가 클래스를 따로 쓰고있는데 이것을 계속 나눠둘 이유가 있을까요??

Yo0oN replied:
1. 네~
2. 회원가입시에 Host와 Member의 화면을 다르게 줄 생각이어서 나눠 사용하였는데 컨트롤러를 나누지 않아도 화면을 다르게 줄 수 있으니 나누지 않아도 됩니다. 둘을 합쳐서 사용하겠습니다

f-lab-dev commented on src/main/java/com/festa/management/SessionManagement.java:
```diff
@@ -19,4 +22,19 @@ public void makeSession(HttpSession httpSession, String sessionKey, Object objec
         httpSession.setAttribute(sessionKey, object);
     }
+
+    public boolean isExistedSession(HttpSession httpSession, String sessionKey){
```
이것은 세션에 인자의 키에 해당하는 데이터가 있는지 확인하는 로직같은데요. **존재하는 세션인가?**라는 메소드명은 이 역할을 잘 나타내고 있지 않은 것 같습니다~

Yo0oN replied:
이부분은 이후 SessionManagement 클래스를 삭제하며 제거하였습니다

f-lab-dev commented on src/main/java/com/festa/management/SessionManagement.java:
```diff
@@ -19,4 +22,19 @@ public void makeSession(HttpSession httpSession, String sessionKey, Object objec
+     * @param httpSession
+     * @param sessionKey
+     */
+    public void deleteSession(HttpSession httpSession, String sessionKey){
```
이것은 그냥 단순히 `httpSession.removeAttribute(sessionKey);`를 감싸고 있는 것 같은데 메소드가 필요한 이유가 있을까요?

Yo0oN replied:
이부분은 이후 SessionManagement 클래스를 삭제하며 제거하였습니다

f-lab-dev commented on src/main/java/com/festa/management/SessionManagement.java:
```diff
@@ -4,6 +4,9 @@
+ * Session을 관리하는 클래스로, 생성, 삭제, 존재 여부 등 관한 것들을 관리하는 클래스
+ */
 @Component
 public class SessionManagement {
```
이 클래스 자체는 세션의 기능을 간단하게 감싸는 기능을 갖고있는 것 같은데 이 클래스의 의도가 궁금합니다~

Yo0oN replied:
해당 프로젝트에 맞게 세션을 따로 관리하는 클래스를 만들어서 부가적인 기능들을 더 추가하고싶었는데 생각보다 기능이 많이 없어서 없애려는 중입니다ㅠㅠ

f-lab-dev commented on src/main/java/com/festa/service/AccountsService.java:
```diff
@@ -62,13 +62,29 @@ public void login(LoginDTO loginDTO, HttpSession httpSession){
+        if (isLogin){
+            sessionManagement.deleteSession(httpSession, loginSessionKey);
+        } else {
+            throw new IllegalStateException("로그인 되어있지 않은 사용자입니다.");
```
로그인이 되어있지 않아도 굳이 예외를 던져줄 필요는 없어보입니다~ 어차피 비운 상태에서 비우는걸 시도해도 사용자입장에선 의도와 다른 동작이 일어나진 않을테니까요

Yo0oN replied:
네~

Yo0oN committed:
[#11]feat: SessionManagement, Controller 수정

1. SessionManagement 삭제
2. HostController 삭제
3. logout 반환값 수정
4. 메서드명 변경

[#11] 로그아웃 기능

f-lab-dev commented on src/main/java/com/festa/service/AccountsService.java:
```diff
@@ -46,14 +42,6 @@ public void signUp(SignUpDTO signUpDTO) {
-     * @param loginDTO
-     * @return
-     */
     public void login(LoginDTO loginDTO, HttpSession httpSession){
```
띄어쓰기 하나 넣으면 좋을 것 같아요~

```suggestion
    public void login(LoginDTO loginDTO, HttpSession httpSession) {
```

Yo0oN replied:
네~

f-lab-dev commented on src/main/java/com/festa/service/AccountsService.java:
```diff
@@ -62,27 +50,15 @@ public void login(LoginDTO loginDTO, HttpSession httpSession){
         LoginDTO resultLoginDTO = accountsDAO.getUserInfoForLogin(loginDTO);
 
-        sessionManagement.makeSession(httpSession, loginSessionKey, resultLoginDTO);
+        httpSession.setAttribute(loginSessionKey, resultLoginDTO);
```
이 클래스는 회원의 정보를 관리하는 클래스 같은데요. 로그인 같은 인증과 관련된 로직은 따로 클래스를 분리하면 어떨까요?

Yo0oN replied:
네! 이후 개발 편의를 위하여 인증 관련된 부분인 로그인, 로그아웃 부분만 클래스를 분리하여 다시 작성하였습니다!

Yo0oN committed:
[#11]feat: 로그인, 로그아웃 관련 클래스 분리

이후 확장시 편리하게 하기 위하여 세션과 같이 인증을 이용하는 부분을 SessionService로 분리

[#11] 로그아웃 기능

f-lab-dev commented on src/main/java/com/festa/common/SessionService.java:
```diff
@@ -0,0 +1,44 @@
+import javax.servlet.http.HttpSession;
+
+@Service
+public class SessionService implements LoginService {
```
`SessionService`라고 하면 그냥 **세션에 대한 뭔가를 하는 서비스**라고 생각하지 로그인과 관련이 있다고는 생각이 들지 않을 것 같습니다~ 좀 더 명확한 이 클래스의 역할에 대한 이름을 지어주세요~

f-lab-dev commented on src/main/java/com/festa/common/SessionService.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Autowired
+    private AccountsDAO accountsDAO;
+    @Autowired
```
여기서는 필드 인젝션을 사용하고 계신데 이유가 있으실까요?

f-lab-dev commented on src/main/java/com/festa/common/SessionService.java:
```diff
@@ -0,0 +1,44 @@
+    @Autowired
+    private AccountsService accountsService;
+
+    private String loginSessionKey = "USER_LOGIN_KEY";
```
이런 값들은 상수로 선언해주시면 좋을 것 같습니다~ 객체의 개수에 따라 이 필드도 함께 새로운 객체가 계속 생성되어서 메모리를 잡아먹을 것 같네요

Yo0oN replied:
네

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -39,13 +42,14 @@ public HttpStatus signUp(@RequestBody @Valid SignUpDTO signUpDTO){
     @PostMapping(value = "/login")
     public HttpStatus login(@RequestBody @Valid LoginDTO loginDTO, HttpSession httpSession){
-        accountsService.login(loginDTO, httpSession);
+        LoginDTO resultLoginDTO = accountsService.getUserInfoLogin(loginDTO);
```
이 PR에서 작성된 클래스들의 의존성을 보자면 `MemberController` -> `AccountService` -> `LoginService` 가 되는데 이 순서가 맞는 순서일까요? 만약 의도를 갖고 설계하신것이라면 설명해주시면 감사하겠습니다~

Yo0oN replied:
아니요!중간에 accountsService 대신 새로 만든 LoginService.getuserInfoLogin을 사용해야 했는데 깜빡하고 지우지 못했나봅니다ㅠㅠ수정하겠습니다!

Yo0oN committed:
[#11]feat:로그아웃 기능

세션을 이용하는 부분 분리 후 로그인과 로그아웃에서 사용 중이던 AccountsService를 사용하는 부분 제거

[#11] 로그아웃 기능

f-lab-dev commented on src/main/java/com/festa/common/SessionService.java:
```diff
@@ -15,14 +15,12 @@
     @Override
     public void login(HttpSession httpSession, LoginDTO loginDTO) {
-        boolean existedID = accountsService.isExistedID(loginDTO.getUserID());
+        boolean existedID = accountsDAO.existedID(loginDTO.getUserID());
```
여기서는 `accountDAO`를 직접 사용하는게 아니라 `accountsService`를 사용하는게 어떨까요?
`accountDAO`는 순수하게 데이터를 불러오는 역할이고, 추가적인 비지니스 로직을 입히기 위해 `accountsService`를 사용하는 것인데 이런 구조로 가면 유저의 정보에 대한 추가적인 비지니스 로직이 `SessionService`로 들어가게 됩니다.

그리고 `SessionService`에 대한 이름도 고민해서 바꿔주세요~

Yo0oN replied:
네!세션을 이용한 로그인 관련 작업을 하고 있기 때문에 SessionLoginService으로 수정하였습니다.

Yo0oN committed:
[#11]feat:로그아웃 기능

SessionService 대신 로그인 관련 작업이라는 것을 명확히 알 수 있도록 SessionLoginService로 수정

[#11] 로그아웃 기능
